src/evm.zig

```diff
@@ -473,6 +473,50 @@ pub fn Evm(config: EvmConfig) type {
                 if (gas_limit < initial_gas_cost) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
                 if (gas_limit < floor_gas_cost) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
             }
+
+            // Validate and calculate blob gas costs for EIP-4844 transactions
+            const blob_count = self.get_blob_versioned_hashes().len;
+            const blob_base_fee = self.get_blob_base_fee();
+            const max_fee_per_blob_gas = self.get_max_fee_per_blob_gas();
+
+            if (!config.eips.validate_blob_gas(
+                blob_count,
+                max_fee_per_blob_gas,
+                blob_base_fee,
+            )) {
+                if (blob_count > primitives.Blob.MAX_BLOBS_PER_TRANSACTION) {
+                    log.err("Too many blobs: {d}, max allowed: {d}", .{ blob_count, primitives.Blob.MAX_BLOBS_PER_TRANSACTION });
+                } else if (blob_count > 0 and max_fee_per_blob_gas == 0) {
+                    log.err("Blob transaction requires max_fee_per_blob_gas to be set", .{});
+                } else if (blob_count > 0 and max_fee_per_blob_gas < blob_base_fee) {
+                    log.err("Max fee per blob gas too low: max={d}, current={d}", .{ self.context.max_fee_per_blob_gas, self.block_info.blob_base_fee });
+                }
+                return CallResult.failure(0);
+            }
+
+            const blob_gas_cost = config.eips.blob_gas_cost(blob_base_fee, blob_count);
+            const max_blob_gas_cost = config.eips.max_blob_gas_cost(max_fee_per_blob_gas, blob_count);
+
+            // Check if sender can afford gas + value + blob gas
+            // We need to check this before execution so we can revert the transaction without spending gas
+            // if the transaction is invalid
+            if (comptime !config.disable_balance_checks) {
```

Comment:
this looks like a bug to me seems like if we disable balance checks like this later we can underflow